### PR TITLE
'Card grid automatic' paragraph, to embed a filtered view. DDFFORM-127

### PR DESCRIPTION
**Pull request by rasben**
~~Requires this `dpl-design` PR to be merged: https://github.com/danskernesdigitalebibliotek/dpl-design-system/pull/439~~

----

This is an addition to the 'card grid manual' paragraph, where the editor chooses 1-6 contents. This paragraph allows the editor to choose a content type, to display the latest 6 items. In the future, we will also introduce a tagging system.

Ontop of that, the view and preprocess functionality has been made general enough to easily add other view styles and filters.

#### Link to issue

https://reload.atlassian.net/browse/DDFFORM-127

------

This builds ontop of the manual version, done here: https://github.com/danskernesdigitalebibliotek/dpl-cms/pull/657

-----

Screenshot 2024-01-18 at 21 07 07

### CODE REVIEW

File: config/sync/views.view.content_paragraphs.yml [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  config:
+    - core.entity_view_mode.node.card
+  module:
+    - node
+    - user
+id: content_paragraphs
+label: 'Content paragraphs'
+module: views
+description: ''
+tag: ''
```
Comment (kasperg) [inline config/sync/views.view.content_paragraphs.yml:14]: Consider adding more information here.

To me this title and description is fairly non-descript but it is heavily tied to the `dpl_filter_paragraphs` module. Can we update it to make this relationship more clear?

File: web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  $variables['#cache']['tags'][] = 'node_list';
+
+  $paragraph = $variables['paragraph'] ?? NULL;
+
+  if (!($paragraph instanceof Paragraph)) {
+    return $variables;
+  }
+
+  $view = Views::getView('content_paragraphs');
+  if (!is_object($view)) {
```
Comment (kasperg) [inline web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module:47]: Please check against the `Drupal\views\ViewExecutable` class.

This is one of the return types for `Views::getView()` and follows the pattern used above and elsewhere in the code.

File: web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    $field = $paragraph->get($filter_name);
+    $filter_value = $field->getString();
+    $filter_value = !empty($filter_value) ? $filter_value : 'all';
+
+    // getString() gets values as comma seperated.
+    // We want it to be seperated with + instead, to show view that it's
+    // OR instead of AND.
+    $filter_value = str_replace(', ', '+', $filter_value);
```
Comment (kasperg) [inline web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module:80]: :+1: Excellent commenting.

File: web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  // as contextual values for the view.
+  foreach ($filters as $filter_name) {
+    // If the paragraph has no matching field, we'll just set the value to
+    // all. THIS IS IMPORTANT, as otherwise there will be a mismatch between
+    // the contextual filters and the keys.
+    if (!$paragraph->hasField($filter_name)) {
+      $arguments[] = 'all';
+
+      continue;
+    }
```
Comment (kasperg) [inline web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module:71]: :+1: Very helpful.

File: web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+ * @return array<mixed>
+ *   See the $variables in theme_preprocess_paragraph__NAME().
+ */
+function _dpl_filter_paragraphs_prepare_filter_view(array $variables, string $item_view_mode): array {
+  // In the preview (AKA backend view), we don't want to load in the whole view.
+  if ($variables['view_mode'] === 'preview') {
+    return $variables;
+  }
+
+  $variables['#cache']['tags'][] = 'node_list';
```
Comment (kasperg) [inline web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module:38]: I wonder why this is necessary. I would expect the child element with the view to have the necessary cache tags including `node_list` and for these to bubble up. 🤔

File: web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+  // Argument = Contextual filters.
+  $view->setArguments($arguments);
+
+  // Making a unique cache key, based on the chosen arguments.
+  $cache_key = "filter_paragraphs_{$item_view_mode}:" . md5(serialize($arguments));
+
+  $view->element['#cache'] = NestedArray::mergeDeep($view->element['#cache'], [
+    'keys' => [$cache_key],
+  ]);
```
Comment (kasperg) [inline web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module:93]: Consider explaining *why* we need to add a unique cache key here.

The code itself is pretty straightforward.

File: config/sync/field.storage.paragraph.field_filter_content_types.yml
Comment (kasperg) [inline config/sync/field.storage.paragraph.field_filter_content_types.yml:1]: Consider if this is a temporary solution.

The field only seems to support node types. What do we do about event series (or instances)? 

Monosnap Edit Article Jesper Stein vinder Læsernes Bogpris for Rampen | DPL CMS | Logged in 2024-01-22 09-19-15

**Review by kasperg — CHANGES_REQUESTED** [review]
:+1: Nice work

I think a few things needs to be considered here. The most important consideration is in regard to selecting filters.

File: web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module (hunk 5 above)
Comment (rasben) [reply web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module:38]: I'll add a code comment to clarifiy :)

File: web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module (hunk 6 above)
Comment (rasben) [reply web/modules/custom/dpl_filter_paragraphs/dpl_filter_paragraphs.module:93]: I'll add a code comment to clarifiy :)

File: config/sync/field.storage.paragraph.field_filter_content_types.yml
Comment (rasben) [reply config/sync/field.storage.paragraph.field_filter_content_types.yml:1]: It turned out to be pretty complicated, and I dont think they ever want to use the card grids for events, so I've moved it into a seperate ticket:

https://reload.atlassian.net/browse/DDFFORM-179

Seeing as we'll need it when we wanna do automatic sliders :)

**Review by kasperg — APPROVED** [review]
:+1: LGTM

**Pull request merged**